### PR TITLE
Refactor backend loading to also check absolute paths

### PR DESCRIPTION
**Pull request by VivekPanyam**
### Summary:
This PR adds some comments to the backend loading system and refactors the loading behavior to support default backends installed at absolute paths.

This helps us start to solve a few problems:
- There are a growing number of interfaces to Neuropod (C++, Python, C, Java, etc) and we don't want unique ways of installing backends per language.
- There are also issues with Java and `LD_LIBRARY_PATH` on macOS because of system integrity protection. This makes it quite difficult to use `LD_LIBRARY_PATH` to add backends for Neuropod to dlopen
-  Models can't choose the version of a framework to use. This is because all the backend `so` files for a type have the same name regardless of the version (so the backend loading system cannot load a specific version). Even if they had different names based on version (e.g. `libneuropod_torchscript_1_4_0_cpu_backend.so`), we would still see issues when adding their containing folders to the library loader search path. This is because then there would be conflicting versions of other unversioned shared objects included with the backends (e.g. `libc10.so` from torch 1.1.0 and `libc10.so` from torch 1.4.0)

We start to solve these by allowing users to install backends at a "well known" path on the system. This way, all installed backends will be available from all languages without any additional configuration.

For now, the "well known" location is `/usr/local/lib/neuropod/`, but this may change or become user-configurable in the future.

We solve the last problem mentioned above by installing each backend in a versioned location (that is not on the system's library path). For example, `/usr/local/lib/neuropod/0.2.0/backends/torchscript_1.4.0_gpu/libneuropod_torchscript_backend.so`.

This means each backend version can now be independently loaded without being on the system library path. See https://github.com/uber/neuropod/issues/348 for more details. 

### Test Plan:
CI + Local testing

Will explore tests with multiple framework versions in CI

### CODE REVIEW

**Review by sepo-eng — APPROVED** [review]
LGTM.

The get_default_backend_map seems like a good candidate for some unit tests.

File: source/neuropod/internal/backend_registration.cc [hunk 1]
```diff
@@ -51,44 +53,74 @@ void init_registrar_if_needed()
...
+
+                for (const auto &version : framework.versions)
+                {
+                    BackendLoadSpec item;
+                    item.type    = framework.type;
+                    item.version = version;
+
+                    if (is_absolute_path)
+                    {
+                        // Ex:
```
Comment (sepo-eng) [inline source/neuropod/internal/backend_registration.cc:99]: Is it worth validating the existence of the file if is_absolute_path is true, and skip appending to 'out' if we find it doesn't exist?

Comment (VivekPanyam) [reply source/neuropod/internal/backend_registration.cc:99]: Probably not for the following reasons:
- If a file exists, it doesn't necessarily mean `dlopen` will work
- For relative paths, internally `dlopen` effectively just tries a bunch of absolute paths based on the system loader paths and `LD_LIBRARY_PATH`. Therefore, the approach here of trying several candidate locations likely isn't going to be "expensive"

**Comment by VivekPanyam** [comment]
> The get_default_backend_map seems like a good candidate for some unit tests.

I have a few PRs in progress related to library loading + changes in CI to make dynamic loading tests easier to build.

Not sure how much value we get out of testing `get_default_backend_map` directly in its current form. Its purpose was to avoid writing out the (fairly large) backend list manually in the code. We could spot check the output of the function, but I'd rather have "exhaustive" multi-version library loading tests in CI to ensure we can actually load all of these backends correctly.

Maybe it makes sense to refactor `get_default_backend_map` to take in a `std::vector<FrameworkInfo>` instead of using a fixed hardcoded one

**Pull request merged**